App Launcher: handle python app terminal links in `positron-run-app` extension

sharon-wang opened the pull request:
### Summary

- addresses #5431

#### Implementation

- we now register a TerminalLinkProvider in the positron-run-app extension which handles the python app urls in the Terminal, instead of going through the usual uri opener flow (which does not have access to the proxied app urls)
- api.ts: we now keep track of the app servers and corresponding terminals so that we can handle the terminal links and resolve them to the correct proxied app url
- reorganized constants, types, api-utils into separate files to reduce the size of the main app launcher api.ts file
- we don't proxy web apps through the positron-proxy when running on Desktop anymore -- this seemed unnecessary 🤷 

#### Demo

https://github.com/user-attachments/assets/7350d6f4-2b9c-442a-b422-8eab797fccc8

#### `positron.appLauncher.terminalAppUrlOpenLocation` setting

This setting allows the user to control where the app url will open when Cmd/Ctrl-clicked in the Terminal. A gear icon shows in the default quickpick that shows when the user Cmd/Ctrl-clicks the terminal link which navigates the user to the setting in the UI.

### Release Notes

#### Bug Fixes

- App Launcher: app urls from Terminal links are now resolved correctly (#5431)

### QA Notes

@:web @:apps @:win

This was most obviously broken for Dash apps on Web, however all supported App Launcher app types should work.

1. Run a supported App Launcher Python app in the Terminal via the play button
2. Note the url that displays in the Viewer when the app is loaded
3. Cmd/Ctrl+click the url of the app in the Terminal
4. A quickpick should open and the url in the quickpick placeholder area should match the URL in the Viewer
5. Selecting either of the options (viewer pane or browser window) should load the app successfully

sharon-wang committed:
handle app launcher terminal links in extension

instead of relying on the default opener behaviour, which is unaware of the proxy servers, handle the terminal links for the app urls directly in the extension.

sharon-wang committed:
cleanup and reorg

sharon-wang commented on extensions/positron-run-app/src/api.ts:
```diff
@@ -389,249 +405,128 @@ export class PositronRunAppApiImpl implements PositronRunApp, vscode.Disposable
-			vscode.l10n.t('Shell integration is now enabled. Would you like to rerun the application?'),
-			rerunApplication,
-			notNow,
+		// Record the known app server URL to proxy server mapping.
+		this.addAppServer(
+			localBaseUri.toString(),
+			options.terminalPid,
+			previewUri,
```
this is the only real change to this `previewUrlInExecutionOutput` method. The rest of the diff is from moving `previewUrlInExecutionOutput` from a standalone function in this file to a method within the `PositronRunAppApiImpl` class

jmcphers commented on extensions/positron-run-app/src/terminalLinkProvider.ts:
```diff
@@ -0,0 +1,82 @@
+		const uri = await vscode.env.asExternalUri(link.proxyUri);
+
+		// Ask the user if they want to open the URL in the Viewer or in the a new browser tab.
+		const viewerPane = vscode.l10n.t('Open in Viewer pane');
```
Nit: It would be nice to not have to answer this question every time a link is opened -- could we remember the user's choice in some way?

jmcphers reviewed (dismissed):
LGTM and does what it says on the tin!

sharon-wang committed:
add config preference for where to open terminal app urls

`positron.appLauncher.terminalAppUrlOpenLocation` defaults to `ask` (show quickpick to ask where to open the URL), `viewer` (open in viewer without asking), or `browser` (open in browser without asking).
